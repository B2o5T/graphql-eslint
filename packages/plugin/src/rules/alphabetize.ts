import {
  Kind,
  ObjectTypeDefinitionNode,
  ObjectTypeExtensionNode,
  InterfaceTypeDefinitionNode,
  InterfaceTypeExtensionNode,
  InputObjectTypeDefinitionNode,
  InputObjectTypeExtensionNode,
  FieldDefinitionNode,
  InputValueDefinitionNode,
  EnumTypeDefinitionNode,
  EnumTypeExtensionNode,
  EnumValueDefinitionNode,
  DirectiveDefinitionNode,
  ArgumentNode,
  VariableDefinitionNode,
  OperationDefinitionNode,
  FieldNode,
  DirectiveNode,
  SelectionSetNode,
  FragmentSpreadNode,
} from 'graphql';
import type { SourceLocation, Comment } from 'estree';
import { GraphQLESLintRule } from '../types';
import { GraphQLESTreeNode } from '../estree-parser';
import { GraphQLESLintRuleListener } from '../testkit';
import { getLocation } from '../utils';

const ALPHABETIZE = 'ALPHABETIZE';

const fieldsEnum: ('ObjectTypeDefinition' | 'InterfaceTypeDefinition' | 'InputObjectTypeDefinition')[] = [
  Kind.OBJECT_TYPE_DEFINITION,
  Kind.INTERFACE_TYPE_DEFINITION,
  Kind.INPUT_OBJECT_TYPE_DEFINITION,
];
const valuesEnum: ['EnumTypeDefinition'] = [Kind.ENUM_TYPE_DEFINITION];
const selectionsEnum: ('OperationDefinition' | 'FragmentDefinition')[] = [
  Kind.OPERATION_DEFINITION,
  Kind.FRAGMENT_DEFINITION,
];
const variablesEnum: ['OperationDefinition'] = [Kind.OPERATION_DEFINITION];
const argumentsEnum: ('FieldDefinition' | 'Field' | 'DirectiveDefinition' | 'Directive')[] = [
  Kind.FIELD_DEFINITION,
  Kind.FIELD,
  Kind.DIRECTIVE_DEFINITION,
  Kind.DIRECTIVE,
];

export type AlphabetizeConfig = {
  fields?: typeof fieldsEnum;
  values?: typeof valuesEnum;
  selections?: typeof selectionsEnum;
  variables?: typeof variablesEnum;
  arguments?: typeof argumentsEnum;
};

const rule: GraphQLESLintRule<[AlphabetizeConfig]> = {
  meta: {
    type: 'suggestion',
    fixable: 'code',
    docs: {
      category: ['Schema', 'Operations'],
      description: [
        'Enforce arrange in alphabetical order for type fields, enum values, input object fields, operation selections and more.',
        '> Note: this rule will autofix only fields without comments (between or around)',
      ].join('\n\n'),
      url: 'https://github.com/dotansimha/graphql-eslint/blob/master/docs/rules/alphabetize.md',
      examples: [
        {
          title: 'Incorrect',
          usage: [{ fields: [Kind.OBJECT_TYPE_DEFINITION] }],
          code: /* GraphQL */ `
            type User {
              password: String
              firstName: String! # should be before "password"
              age: Int # should be before "firstName"
              lastName: String!
            }
          `,
        },
        {
          title: 'Correct',
          usage: [{ fields: [Kind.OBJECT_TYPE_DEFINITION] }],
          code: /* GraphQL */ `
            type User {
              age: Int
              firstName: String!
              lastName: String!
              password: String
            }
          `,
        },
        {
          title: 'Incorrect',
          usage: [{ values: [Kind.ENUM_TYPE_DEFINITION] }],
          code: /* GraphQL */ `
            enum Role {
              SUPER_ADMIN
              ADMIN # should be before "SUPER_ADMIN"
              USER
              GOD # should be before "USER"
            }
          `,
        },
        {
          title: 'Correct',
          usage: [{ values: [Kind.ENUM_TYPE_DEFINITION] }],
          code: /* GraphQL */ `
            enum Role {
              ADMIN
              GOD
              SUPER_ADMIN
              USER
            }
          `,
        },
        {
          title: 'Incorrect',
          usage: [{ selections: [Kind.OPERATION_DEFINITION] }],
          code: /* GraphQL */ `
            query {
              me {
                firstName
                lastName
                email # should be before "lastName"
              }
            }
          `,
        },
        {
          title: 'Correct',
          usage: [{ selections: [Kind.OPERATION_DEFINITION] }],
          code: /* GraphQL */ `
            query {
              me {
                email
                firstName
                lastName
              }
            }
          `,
        },
      ],
      configOptions: {
        schema: [
          {
            fields: fieldsEnum,
            values: valuesEnum,
            arguments: argumentsEnum,
          },
        ],
        operations: [
          {
            selections: selectionsEnum,
            variables: variablesEnum,
            arguments: [Kind.FIELD, Kind.DIRECTIVE],
          },
        ],
      },
    },
    messages: {
      [ALPHABETIZE]: '"{{ currName }}" should be before "{{ prevName }}"',
    },
    schema: {
      type: 'array',
      minItems: 1,
      maxItems: 1,
      items: {
        type: 'object',
        additionalProperties: false,
        minProperties: 1,
        properties: {
          fields: {
            type: 'array',
            uniqueItems: true,
            minItems: 1,
            items: {
              enum: fieldsEnum,
            },
            description: 'Fields of `type`, `interface`, and `input`',
          },
          values: {
            type: 'array',
            uniqueItems: true,
            minItems: 1,
            items: {
              enum: valuesEnum,
            },
            description: 'Values of `enum`',
          },
          selections: {
            type: 'array',
            uniqueItems: true,
            minItems: 1,
            items: {
              enum: selectionsEnum,
            },
            description: 'Selections of operations (`query`, `mutation` and `subscription`) and `fragment`',
          },
          variables: {
            type: 'array',
            uniqueItems: true,
            minItems: 1,
            items: {
              enum: variablesEnum,
            },
            description: 'Variables of operations (`query`, `mutation` and `subscription`)',
          },
          arguments: {
            type: 'array',
            uniqueItems: true,
            minItems: 1,
            items: {
              enum: argumentsEnum,
            },
            description: 'Arguments of fields and directives',
          },
        },
      },
    },
  },
  create(context) {
    function isOnSameLineNodeAndComment(beforeNode: { loc: SourceLocation }, afterNode: Comment): boolean {
      return beforeNode.loc.end.line === afterNode.loc.start.line;
    }

    function checkNodes(
      nodes: GraphQLESTreeNode<
        | FieldDefinitionNode
        | InputValueDefinitionNode
        | EnumValueDefinitionNode
        | FieldNode
        | FragmentSpreadNode
        | ArgumentNode
        | VariableDefinitionNode['variable']
      >[]
    ) {
      // Starts from 1, ignore nodes.length <= 1
      for (let i = 1; i < nodes.length; i += 1) {
        const prevNode = nodes[i - 1];
        const currNode = nodes[i];
        const prevName = prevNode.name.value;
        const currName = currNode.name.value;
        if (prevName.localeCompare(currName) === 1) {
          const isVariableNode = currNode.kind === Kind.VARIABLE;

          context.report({
            loc: getLocation(currNode.loc, currName, { offsetEnd: isVariableNode ? 0 : 1 }),
            messageId: ALPHABETIZE,
            data: isVariableNode
              ? {
                  currName: `$${currName}`,
                  prevName: `$${prevName}`,
                }
              : { currName, prevName },
            *fix(fixer) {
              const prev = prevNode as any;
              const curr = currNode as any;
              const sourceCode = context.getSourceCode();

              const beforeComments = sourceCode.getCommentsBefore(prev);
              if (beforeComments.length > 0) {
                const tokenBefore = sourceCode.getTokenBefore(prev);
                const lastBeforeComment = beforeComments.at(-1);
                if (!tokenBefore || !isOnSameLineNodeAndComment(tokenBefore, lastBeforeComment)) return;
              }

              const betweenComments = sourceCode.getCommentsBefore(curr);
              if (betweenComments.length > 0) return;

              const [firstAfterComment] = sourceCode.getCommentsAfter(curr);
              if (firstAfterComment && isOnSameLineNodeAndComment(curr, firstAfterComment)) return;

              yield fixer.replaceText(prev, sourceCode.getText(curr));
              yield fixer.replaceText(curr, sourceCode.getText(prev));
            },
          });
        }
      }
    }

    const opts = context.options[0];
    const fields = new Set(opts.fields ?? []);
    const listeners: GraphQLESLintRuleListener = {};

    const kinds = [
      fields.has(Kind.OBJECT_TYPE_DEFINITION) && [Kind.OBJECT_TYPE_DEFINITION, Kind.OBJECT_TYPE_EXTENSION],
      fields.has(Kind.INTERFACE_TYPE_DEFINITION) && [Kind.INTERFACE_TYPE_DEFINITION, Kind.INTERFACE_TYPE_EXTENSION],
      fields.has(Kind.INPUT_OBJECT_TYPE_DEFINITION) && [
        Kind.INPUT_OBJECT_TYPE_DEFINITION,
        Kind.INPUT_OBJECT_TYPE_EXTENSION,
      ],
    ]
      .filter(Boolean)
      .flat();

    const fieldsSelector = kinds.join(',');

    const hasEnumValues = opts.values?.[0] === Kind.ENUM_TYPE_DEFINITION;
    const selectionsSelector = opts.selections?.join(',');
    const hasVariables = opts.variables?.[0] === Kind.OPERATION_DEFINITION;
    const argumentsSelector = opts.arguments?.join(',');

    if (fieldsSelector) {
      listeners[fieldsSelector] = (
        node: GraphQLESTreeNode<
          | ObjectTypeDefinitionNode
          | ObjectTypeExtensionNode
          | InterfaceTypeDefinitionNode
          | InterfaceTypeExtensionNode
          | InputObjectTypeDefinitionNode
          | InputObjectTypeExtensionNode
        >
      ) => {
        checkNodes(node.fields);
      };
    }

    if (hasEnumValues) {
      const enumValuesSelector = [Kind.ENUM_TYPE_DEFINITION, Kind.ENUM_TYPE_EXTENSION].join(',');
      listeners[enumValuesSelector] = (node: GraphQLESTreeNode<EnumTypeDefinitionNode | EnumTypeExtensionNode>) => {
        checkNodes(node.values);
      };
    }

    if (selectionsSelector) {
      listeners[`:matches(${selectionsSelector}) SelectionSet`] = (node: GraphQLESTreeNode<SelectionSetNode>) => {
        checkNodes(
          node.selections
            // inline fragment don't have name, so we skip them
            .filter(selection => selection.kind !== Kind.INLINE_FRAGMENT)
            .map(selection =>
              // sort by alias is field is renamed
              'alias' in selection && selection.alias ? ({ name: selection.alias } as any) : selection
            )
        );
      };
    }

    if (hasVariables) {
      listeners.OperationDefinition = (node: GraphQLESTreeNode<OperationDefinitionNode>) => {
        checkNodes(node.variableDefinitions.map(varDef => varDef.variable));
      };
    }

    if (argumentsSelector) {
      listeners[argumentsSelector] = (
        node: GraphQLESTreeNode<FieldDefinitionNode | FieldNode | DirectiveDefinitionNode | DirectiveNode>
      ) => {
        checkNodes(node.arguments);
      };
    }

    return listeners;
  },
};

export default rule;
