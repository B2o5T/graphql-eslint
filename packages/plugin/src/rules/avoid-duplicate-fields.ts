import { Kind } from 'graphql';
import { GraphQLESLintRule } from '../types';

const AVOID_DUPLICATE_FIELDS = 'AVOID_DUPLICATE_FIELDS';

const ensureUnique = () => {
  const set = new Set<string>();

  return {
    add: (item: string, onError: () => void) => {
      if (set.has(item)) {
        onError();
      } else {
        set.add(item);
      }
    },
  };
};

const rule: GraphQLESLintRule<[], false> = {
  meta: {
    type: 'suggestion',
    docs: {
      requiresSchema: false,
      requiresSiblings: false,
      description:
        'Checks for duplicate fields in selection set, variables in operation definition, or in arguments set of a field.',
      category: 'Stylistic Issues',
      url: 'https://github.com/dotansimha/graphql-eslint/blob/master/docs/rules/avoid-duplicate-fields.md',
      examples: [
        {
          title: 'Incorrect',
          code: /* GraphQL */ `
            query getUserDetails {
              user {
                name # first
                email
                name # second
              }
            }
          `,
        },
        {
          title: 'Incorrect',
          code: /* GraphQL */ `
            query getUsers {
              users(
                first: 100
                skip: 50
                after: "cji629tngfgou0b73kt7vi5jo"
                first: 100 # duplicate argument
              ) {
                id
              }
            }
          `,
        },
        {
          title: 'Incorrect',
          code: /* GraphQL */ `
            query getUsers($first: Int!, $first: Int!) {
              # Duplicate variable
              users(first: 100, skip: 50, after: "cji629tngfgou0b73kt7vi5jo") {
                id
              }
            }
          `,
        },
      ],
    },
    messages: {
      [AVOID_DUPLICATE_FIELDS]: `{{ type }} "{{ fieldName }}" defined multiple times.`,
    },
  },
  create(context) {
    return {
      OperationDefinition(node) {
        const uniqueCheck = ensureUnique();

        for (const arg of node.variableDefinitions || []) {
          uniqueCheck.add(arg.variable.name.value, () => {
            context.report({
              messageId: AVOID_DUPLICATE_FIELDS,
              data: {
                type: 'Operation variable',
                fieldName: arg.variable.name.value,
              },
              node: arg,
            });
          });
        }
      },
      Field(node) {
        const uniqueCheck = ensureUnique();

        for (const arg of node.arguments || []) {
          uniqueCheck.add(arg.name.value, () => {
            context.report({
              messageId: AVOID_DUPLICATE_FIELDS,
              data: {
                type: 'Field argument',
                fieldName: arg.name.value,
              },
              node: arg,
            });
          });
        }
      },
      SelectionSet(node) {
        const uniqueCheck = ensureUnique();

        for (const selection of node.selections || []) {
          if (selection.kind === Kind.FIELD) {
            const nameToCheck = selection.alias?.value || selection.name.value;

            uniqueCheck.add(nameToCheck, () => {
              context.report({
                messageId: AVOID_DUPLICATE_FIELDS,
                data: {
                  type: 'Field',
                  fieldName: nameToCheck,
                },
                node: selection,
              });
            });
          }
        }
      },
    };
  },
};

export default rule;
