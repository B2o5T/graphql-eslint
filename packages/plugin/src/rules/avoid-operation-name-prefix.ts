import { GraphQLESLintRule, GraphQLESlintRuleContext } from '../types';
import { GraphQLESTreeNode } from '../estree-parser/estree-ast';
import { OperationDefinitionNode, FragmentDefinitionNode } from 'graphql';

type AvoidOperationNamePrefixConfig = {
  keywords: string[];
  caseSensitive?: boolean;
};

const AVOID_OPERATION_NAME_PREFIX = 'AVOID_OPERATION_NAME_PREFIX';

function verifyRule(
  context: GraphQLESlintRuleContext<AvoidOperationNamePrefixConfig>,
  node: GraphQLESTreeNode<OperationDefinitionNode> | GraphQLESTreeNode<FragmentDefinitionNode>
) {
  const caseSensitive = context.options[0].caseSensitive;

  if (node && node.name && node.name.value !== '') {
    for (const keyword of context.options[0].keywords) {
      const testKeyword = caseSensitive ? keyword : keyword.toLowerCase();
      const testName = caseSensitive ? node.name.value : node.name.value.toLowerCase();

      if (testName.startsWith(testKeyword)) {
        context.report({
          loc: {
            start: {
              line: node.name.loc.start.line,
              column: node.name.loc.start.column - 1,
            },
            end: {
              line: node.name.loc.start.line,
              column: node.name.loc.start.column + testKeyword.length - 1,
            },
          },
          data: {
            invalidPrefix: keyword,
          },
          messageId: AVOID_OPERATION_NAME_PREFIX,
        });
      }
    }
  }
}

const rule: GraphQLESLintRule<AvoidOperationNamePrefixConfig> = {
  meta: {
    messages: {
      [AVOID_OPERATION_NAME_PREFIX]: `Forbidden operation name prefix: "{{ invalidPrefix }}"`,
    },
    schema: {
      type: 'array',
      minItems: 1,
      maxItems: 1,
      additionalItems: false,
      items: {
        additionalProperties: false,
        type: 'object',
        required: ['keywords'],
        properties: {
          caseSensitive: {
            default: false,
            type: 'boolean',
          },
          keywords: {
            additionalItems: false,
            type: 'array',
            minItems: 1,
            items: {
              type: 'string',
            },
          },
        },
      },
    },
  },
  create(context) {
    return {
      OperationDefinition(node) {
        verifyRule(context, node);
      },
      FragmentDefinition(node) {
        verifyRule(context, node);
      },
    };
  },
};

export default rule;
