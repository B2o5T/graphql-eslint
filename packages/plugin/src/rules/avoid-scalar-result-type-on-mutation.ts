import { Kind, FieldDefinitionNode, isScalarType } from 'graphql';
import { requireGraphQLSchemaFromContext, getTypeName } from '../utils';
import { GraphQLESLintRule } from '../types';
import { GraphQLESTreeNode } from '../estree-parser';

const rule: GraphQLESLintRule = {
  meta: {
    type: 'suggestion',
    docs: {
      category: 'Best Practices',
      description: 'Avoid scalar result type on mutation type to make sure to return a valid state.',
      url: 'https://github.com/dotansimha/graphql-eslint/blob/master/docs/rules/avoid-scalar-result-type-on-mutation.md',
      requiresSchema: true,
      examples: [
        {
          title: 'Incorrect',
          code: /* GraphQL */ `
            type Mutation {
              createUser: Boolean
            }
          `,
        },
        {
          title: 'Correct',
          code: /* GraphQL */ `
            type Mutation {
              createUser: User!
            }
          `,
        },
      ],
    },
  },
  create(context) {
    const schema = requireGraphQLSchemaFromContext('avoid-scalar-result-type-on-mutation', context);
    const mutationType = schema.getMutationType();
    if (!mutationType) {
      return {};
    }
    const selector = `:matches(${Kind.OBJECT_TYPE_DEFINITION}, ${Kind.OBJECT_TYPE_EXTENSION})[name.value=${mutationType.name}] > ${Kind.FIELD_DEFINITION}`;

    return {
      [selector](node: GraphQLESTreeNode<FieldDefinitionNode>) {
        const rawNode = node.rawNode();
        const typeName = getTypeName(rawNode);
        const graphQLType = schema.getType(typeName);
        if (isScalarType(graphQLType)) {
          context.report({
            node,
            message: `Unexpected scalar result type "${typeName}"`,
          });
        }
      },
    };
  },
};

export default rule;
