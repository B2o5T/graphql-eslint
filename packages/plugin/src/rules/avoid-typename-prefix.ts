import {
  InterfaceTypeDefinitionNode,
  InterfaceTypeExtensionNode,
  ObjectTypeDefinitionNode,
  ObjectTypeExtensionNode,
} from 'graphql';
import { GraphQLESTreeNode } from '../estree-parser';
import { GraphQLESLintRule } from '../types';
import { getLocation } from '../utils';

const AVOID_TYPENAME_PREFIX = 'AVOID_TYPENAME_PREFIX';

const rule: GraphQLESLintRule = {
  meta: {
    type: 'suggestion',
    docs: {
      category: 'Best Practices',
      description: 'Enforces users to avoid using the type name in a field name while defining your schema.',
      recommended: true,
      url: 'https://github.com/dotansimha/graphql-eslint/blob/master/docs/rules/avoid-typename-prefix.md',
      examples: [
        {
          title: 'Incorrect',
          code: /* GraphQL */ `
            type User {
              userId: ID!
            }
          `,
        },
        {
          title: 'Correct',
          code: /* GraphQL */ `
            type User {
              id: ID!
            }
          `,
        },
      ],
    },
    messages: {
      [AVOID_TYPENAME_PREFIX]: `Field "{{ fieldName }}" starts with the name of the parent type "{{ typeName }}"`,
    },
    schema: [],
  },
  create(context) {
    return {
      'ObjectTypeDefinition, ObjectTypeExtension, InterfaceTypeDefinition, InterfaceTypeExtension'(
        node: GraphQLESTreeNode<
          ObjectTypeDefinitionNode | ObjectTypeExtensionNode | InterfaceTypeDefinitionNode | InterfaceTypeExtensionNode
        >
      ) {
        const typeName = node.name.value;
        const lowerTypeName = typeName.toLowerCase();

        for (const field of node.fields) {
          const fieldName = field.name.value;

          if (fieldName.toLowerCase().startsWith(lowerTypeName)) {
            context.report({
              data: {
                fieldName,
                typeName,
              },
              messageId: AVOID_TYPENAME_PREFIX,
              loc: getLocation(field.loc, lowerTypeName),
            });
          }
        }
      },
    };
  },
};

export default rule;
