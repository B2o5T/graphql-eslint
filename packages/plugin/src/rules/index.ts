/*
 * ðŸš¨ IMPORTANT! Do not manually modify this file. Run: `pnpm generate:configs`
 */

import { rule as alphabetize } from './alphabetize/index.js';
import { rule as descriptionStyle } from './description-style/index.js';
import { GRAPHQL_JS_VALIDATIONS } from './graphql-js-validation.js';
import { rule as inputName } from './input-name/index.js';
import { rule as loneExecutableDefinition } from './lone-executable-definition/index.js';
import { rule as matchDocumentFilename } from './match-document-filename/index.js';
import { rule as namingConvention } from './naming-convention/index.js';
import { rule as noAnonymousOperations } from './no-anonymous-operations/index.js';
import { rule as noDeprecated } from './no-deprecated/index.js';
import { rule as noDuplicateFields } from './no-duplicate-fields/index.js';
import { rule as noHashtagDescription } from './no-hashtag-description/index.js';
import { rule as noOnePlaceFragments } from './no-one-place-fragments/index.js';
import { rule as noRootType } from './no-root-type/index.js';
import { rule as noScalarResultTypeOnMutation } from './no-scalar-result-type-on-mutation/index.js';
import { rule as noTypenamePrefix } from './no-typename-prefix/index.js';
import { rule as noUnreachableTypes } from './no-unreachable-types/index.js';
import { rule as noUnusedFields } from './no-unused-fields/index.js';
import { rule as relayArguments } from './relay-arguments/index.js';
import { rule as relayConnectionTypes } from './relay-connection-types/index.js';
import { rule as relayEdgeTypes } from './relay-edge-types/index.js';
import { rule as relayPageInfo } from './relay-page-info/index.js';
import { rule as requireDeprecationDate } from './require-deprecation-date/index.js';
import { rule as requireDeprecationReason } from './require-deprecation-reason/index.js';
import { rule as requireDescription } from './require-description/index.js';
import { rule as requireFieldOfTypeQueryInMutationResult } from './require-field-of-type-query-in-mutation-result/index.js';
import { rule as requireImportFragment } from './require-import-fragment/index.js';
import { rule as requireNullableFieldsWithOneof } from './require-nullable-fields-with-oneof/index.js';
import { rule as requireNullableResultInRoot } from './require-nullable-result-in-root/index.js';
import { rule as requireSelections } from './require-selections/index.js';
import { rule as requireTypePatternWithOneof } from './require-type-pattern-with-oneof/index.js';
import { rule as selectionSetDepth } from './selection-set-depth/index.js';
import { rule as strictIdInTypes } from './strict-id-in-types/index.js';
import { rule as uniqueEnumValueNames } from './unique-enum-value-names/index.js';
import { rule as uniqueFragmentName } from './unique-fragment-name/index.js';
import { rule as uniqueOperationName } from './unique-operation-name/index.js';

export const rules = {
  ...GRAPHQL_JS_VALIDATIONS,
  alphabetize,
  'description-style': descriptionStyle,
  'input-name': inputName,
  'lone-executable-definition': loneExecutableDefinition,
  'match-document-filename': matchDocumentFilename,
  'naming-convention': namingConvention,
  'no-anonymous-operations': noAnonymousOperations,
  'no-deprecated': noDeprecated,
  'no-duplicate-fields': noDuplicateFields,
  'no-hashtag-description': noHashtagDescription,
  'no-one-place-fragments': noOnePlaceFragments,
  'no-root-type': noRootType,
  'no-scalar-result-type-on-mutation': noScalarResultTypeOnMutation,
  'no-typename-prefix': noTypenamePrefix,
  'no-unreachable-types': noUnreachableTypes,
  'no-unused-fields': noUnusedFields,
  'relay-arguments': relayArguments,
  'relay-connection-types': relayConnectionTypes,
  'relay-edge-types': relayEdgeTypes,
  'relay-page-info': relayPageInfo,
  'require-deprecation-date': requireDeprecationDate,
  'require-deprecation-reason': requireDeprecationReason,
  'require-description': requireDescription,
  'require-field-of-type-query-in-mutation-result': requireFieldOfTypeQueryInMutationResult,
  'require-import-fragment': requireImportFragment,
  'require-nullable-fields-with-oneof': requireNullableFieldsWithOneof,
  'require-nullable-result-in-root': requireNullableResultInRoot,
  'require-selections': requireSelections,
  'require-type-pattern-with-oneof': requireTypePatternWithOneof,
  'selection-set-depth': selectionSetDepth,
  'strict-id-in-types': strictIdInTypes,
  'unique-enum-value-names': uniqueEnumValueNames,
  'unique-fragment-name': uniqueFragmentName,
  'unique-operation-name': uniqueOperationName,
};
