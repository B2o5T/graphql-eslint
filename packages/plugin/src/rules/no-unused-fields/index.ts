import { FieldDefinitionNode, GraphQLSchema, TypeInfo, visit, visitWithTypeInfo } from 'graphql';
import { GraphQLProjectConfig } from 'graphql-config';
import { FromSchema } from 'json-schema-to-ts';
import { ModuleCache } from '../../cache.js';
import { SiblingOperations } from '../../siblings.js';
import { GraphQLESLintRule, GraphQLESLintRuleListener, GraphQLESTreeNode } from '../../types.js';
import { requireGraphQLSchemaFromContext, requireSiblingsOperations } from '../../utils.js';

const RULE_ID = 'no-unused-fields';

const RELAY_SCHEMA = /* GraphQL */ `
  # Root Query Type
  type Query {
    user: User
  }

  # User Type
  type User {
    id: ID!
    name: String!
    friends(first: Int, after: String): FriendConnection!
  }

  # FriendConnection Type (Relay Connection)
  type FriendConnection {
    edges: [FriendEdge]
    pageInfo: PageInfo!
  }

  # FriendEdge Type
  type FriendEdge {
    cursor: String!
    node: Friend!
  }

  # Friend Type
  type Friend {
    id: ID!
    name: String!
  }

  # PageInfo Type (Relay Pagination)
  type PageInfo {
    hasPreviousPage: Boolean!
    hasNextPage: Boolean!
    startCursor: String
    endCursor: String
  }
`;

const RELAY_QUERY = /* GraphQL */ `
  query {
    user {
      id
      name
      friends(first: 10) {
        edges {
          node {
            id
            name
          }
        }
      }
    }
  }
`;

const schema = {
  type: 'array',
  maxItems: 1,
  items: {
    type: 'object',
    additionalProperties: false,
    properties: {
      ignoredFieldSelectors: {
        type: 'array',
        uniqueItems: true,
        minItems: 1,
        description: [
          'Fields that will be ignored and are allowed to be unused.',
          '',
          '> These fields are defined by ESLint [`selectors`](https://eslint.org/docs/developer-guide/selectors).',
          '> Paste or drop code into the editor in [ASTExplorer](https://astexplorer.net) and inspect the generated AST to compose your selector.',
        ].join('\n'),
        items: {
          type: 'string',
          pattern: '^\\[(.+)]$',
        },
      },
    },
  },
} as const;

export type RuleOptions = FromSchema<typeof schema>;

type UsedFields = Record<string, Set<string>>;

const usedFieldsCache = new ModuleCache<GraphQLProjectConfig['schema'], UsedFields>();

function getUsedFields(schema: GraphQLSchema, operations: SiblingOperations): UsedFields {
  // We don't want cache usedFields on test environment
  // Otherwise usedFields will be same for all tests
  const cachedValue = usedFieldsCache.get(schema);
  if (process.env.NODE_ENV !== 'test' && cachedValue) {
    return cachedValue;
  }
  const usedFields: UsedFields = Object.create(null);
  const typeInfo = new TypeInfo(schema);

  const visitor = visitWithTypeInfo(typeInfo, {
    Field(node): false | void {
      const fieldDef = typeInfo.getFieldDef();
      if (!fieldDef) {
        // skip visiting this node if field is not defined in schema
        return false;
      }
      const parentTypeName = typeInfo.getParentType()!.name;
      const fieldName = node.name.value;

      usedFields[parentTypeName] ??= new Set();
      usedFields[parentTypeName].add(fieldName);
    },
  });

  const allDocuments = [...operations.getOperations(), ...operations.getFragments()];
  for (const { document } of allDocuments) {
    visit(document, visitor);
  }
  usedFieldsCache.set(schema, usedFields);
  return usedFields;
}

const RELAY_DEFAULT_IGNORED_FIELD_SELECTORS = [
  '[parent.name.value=PageInfo][name.value=endCursor]',
  '[parent.name.value=PageInfo][name.value=startCursor]',
  '[parent.name.value=PageInfo][name.value=hasNextPage]',
  '[parent.name.value=PageInfo][name.value=hasPreviousPage]',
  '[parent.name.value=/Edge$/][name.value=cursor]',
  '[parent.name.value=/Connection$/][name.value=pageInfo]',
];

export const rule: GraphQLESLintRule<RuleOptions> = {
  meta: {
    messages: {
      [RULE_ID]: 'Field "{{fieldName}}" is unused',
    },
    docs: {
      description: 'Requires all fields to be used at some level by siblings operations.',
      category: 'Schema',
      url: `https://the-guild.dev/graphql/eslint/rules/${RULE_ID}`,
      requiresSiblings: true,
      requiresSchema: true,
      // Requires documents to be set
      isDisabledForAllConfig: true,
      examples: [
        {
          title: 'Incorrect',
          code: /* GraphQL */ `
            type User {
              id: ID!
              name: String
              someUnusedField: String
            }

            type Query {
              me: User
            }

            query {
              me {
                id
                name
              }
            }
          `,
        },
        {
          title: 'Correct',
          code: /* GraphQL */ `
            type User {
              id: ID!
              name: String
            }

            type Query {
              me: User
            }

            query {
              me {
                id
                name
              }
            }
          `,
        },
        {
          title: 'Correct (ignoring fields)',
          usage: [{ ignoredFieldSelectors: RELAY_DEFAULT_IGNORED_FIELD_SELECTORS }],
          code: /* GraphQL */ `
            # schema
            ${RELAY_SCHEMA}

            # query
            ${RELAY_QUERY}
          `,
        },
      ],
    },
    type: 'suggestion',
    schema,
    hasSuggestions: true,
  },
  create(context) {
    const schema = requireGraphQLSchemaFromContext(RULE_ID, context);
    const siblingsOperations = requireSiblingsOperations(RULE_ID, context);
    const usedFields = getUsedFields(schema, siblingsOperations);
    const { ignoredFieldSelectors } = context.options[0] || {};
    const selector = (ignoredFieldSelectors || []).reduce(
      (acc, selector) => `${acc}:not(${selector})`,
      'FieldDefinition',
    );
    return {
      [selector](node: GraphQLESTreeNode<FieldDefinitionNode>) {
        const fieldName = node.name.value;
        const parentTypeName = node.parent.name.value;
        const isUsed = usedFields[parentTypeName]?.has(fieldName);
        if (isUsed) {
          return;
        }

        context.report({
          node: node.name,
          messageId: RULE_ID,
          data: { fieldName },
          suggest: [
            {
              desc: `Remove \`${fieldName}\` field`,
              fix(fixer) {
                const sourceCode = context.getSourceCode() as any;
                const tokenBefore = sourceCode.getTokenBefore(node);
                const tokenAfter = sourceCode.getTokenAfter(node);
                const isEmptyType = tokenBefore.type === '{' && tokenAfter.type === '}';
                return fixer.remove((isEmptyType ? node.parent : node) as any);
              },
            },
          ],
        });
      },
    };
  },
};
