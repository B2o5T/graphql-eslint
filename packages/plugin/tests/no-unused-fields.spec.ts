import { rule } from '../src/rules/no-unused-fields';

const SCHEMA = /* GraphQL */ `
  type User {
    id: ID!
    firstName: String
    lastName: String
    age: Int
    address: Address
  }

  type Address {
    country: String!
    zip: String!
    events: [Event!]!
  }

  enum EventName {
    CREATE
    UPDATE
    DELETE
  }

  type Event {
    by: User
    name: EventName
    data: String
  }

  type Query {
    user(id: ID!): User
  }

  type Mutation {
    createUser(firstName: String!): User
    deleteUser(id: ID!): User
  }
`;

const ruleTester = new RuleTester({ schema: SCHEMA });

ruleTester.run('no-unused-fields', rule, {
  valid: [
    {
      code: SCHEMA,
      parserOptions: {
        documents: /* GraphQL */ `
          {
            user(id: 1) {
              ... on User {
                address {
                  zip
                  events {
                    ... on Event {
                      by {
                        id
                      }
                      can_rename: name
                      data
                    }
                  }
                }
              }
            }
          }

          fragment UserFields on User {
            can_rename: firstName
            lastName
          }

          mutation {
            deleteUser(id: 2) {
              age
            }
            createUser(firstName: "Foo") {
              address {
                country
              }
            }
          }
        `,
      },
    },
  ],
  invalid: [
    {
      code: /* GraphQL */ `
        type User {
          id: ID!
          firstName: String
        }
      `,
      parserOptions: {
        documents: /* GraphQL */ `
          {
            user(id: 1) {
              id
            }
          }
        `,
      },
      errors: [{ message: 'Field "firstName" is unused' }],
    },
    {
      code: /* GraphQL */ `
        type Query {
          user(id: ID!): User
        }

        type Mutation {
          deleteUser(id: ID!): User
        }
      `,
      parserOptions: {
        documents: /* GraphQL */ `
          {
            user(id: 1) {
              id
            }
          }
        `,
      },
      errors: [{ message: 'Field "deleteUser" is unused' }],
    },
  ],
});
