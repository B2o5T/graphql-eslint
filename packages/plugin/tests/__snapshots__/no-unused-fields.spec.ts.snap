// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
  1 |
  2 |         # normalize graphql
  3 |         type User {
  4 |           id: ID!
> 5 |           firstName: String
    |           ^^^^^^^^^ Field "firstName" is unused
  6 |         }
  7 |       
`;

exports[` 2`] = `
   1 |
   2 |         # normalize graphql
   3 |         type Query {
   4 |           user(id: ID!): User
   5 |         }
   6 |
   7 |         type Mutation {
>  8 |           deleteUser(id: ID!): User
     |           ^^^^^^^^^^ Field "deleteUser" is unused
   9 |         }
  10 |       
`;
